### DISCUSSION TITLE
How does egg scale to keep track of two different programs with shared substructure?

### DISCUSSION THREAD

**Discussion by brando90** [egraphs-good/egg#173] [Q&A]
This might be a very simple question but I was wondering - if I have two different terms but that share substructure, can egg keep track of both together or do I have to instantiate two separate e-graphs? especially in real a real programming language e.g. say I have to programs in python that share help funtions or expression or something like that, how would store it and use it efficiently with egg, is that possible?

A different way to ask is that I noticed that the input for an e-graph is 1 single term (from the videos). But what if I have TWO terms, is it possible do I feed to two to an e-graph?

What if I am building these two terms incrementally, does that make a difference in how an e-graph is built for them?

**Comment by mwillsey** [comment]
You can add as many terms as you want to an e-graph. In fact, rewrites are doing just that all the time! It's just that adding one term, rewriting it a bunch, and extracting the best is a common use case that I use for examples.
